### PR TITLE
RSDK 893 refresh in checkConnection when needed

### PR DESCRIPTION
**Pull request by purplenicole730**

### CODE REVIEW

File: robot/client/client.go [hunk 1]
```diff
@@ -404,8 +414,14 @@ func (rc *RobotClient) checkConnection(ctx context.Context, checkEvery, reconnec
 			rc.Logger().Debugw("successfully reconnected remote at address", "address", rc.address)
 		} else {
 			check := func() error {
-				if _, _, err := rc.resources(ctx); err != nil {
-					return err
+				if refresh == true {
```
Comment (cheukt) [inline robot/client/client.go:417]: linter error here

**Review by cheukt — CHANGES_REQUESTED** [review]
we should also add tests `TestClientRefresh` should be a good base. We want to make sure that we're only calling `ResourceNames` and `ResourceRPCSubtypes` once per refresh cycle

File: robot/client/client.go [hunk 2]
```diff
@@ -254,17 +254,27 @@ func New(ctx context.Context, address string, logger golog.Logger, opts ...Robot
...
+			rc.checkConnection(closeCtx, checkConnectedTime, reconnectTime, refresh)
 		}, rc.activeBackgroundWorkers.Done)
+
+		if refresh == true {
+			return rc, nil
+		}
 	}
 
-	if checkConnectedTime > 0 && reconnectTime > 0 {
+	if refreshTime > 0 {
```
[the remark this answers is not in the document]
Comment (cheukt) [reply robot/client/client.go:274]: nvm, misread

File: robot/client/client_test.go [hunk 3]
```diff
@@ -507,10 +510,15 @@ func TestClientRefresh(t *testing.T) {
...
 	)
+	test.That(t, callCountSubtypes, test.ShouldEqual, 1)
+	test.That(t, callCountNames, test.ShouldEqual, 1)
 	test.That(t, err, test.ShouldBeNil)
 	<-calledEnough
 	test.That(t, time.Since(start), test.ShouldBeGreaterThanOrEqualTo, 5*dur)
 	test.That(t, time.Since(start), test.ShouldBeLessThanOrEqualTo, 10*dur)
+	// during this duration, the client refreshed 4 additional times
+	test.That(t, callCountSubtypes, test.ShouldEqual, 5)
+	test.That(t, callCountNames, test.ShouldEqual, 6)
```
Comment (purplenicole730) [inline robot/client/client_test.go:521]: I'm not entirely sure why `callCountSubtypes` and `callCountNames` are different. I thought they would be the same, but repeated runs of the tests show that these are the values.

File: robot/client/client_test.go [hunk 4]
```diff
@@ -473,231 +473,194 @@ func TestStatusClient(t *testing.T) {
...
+		injectRobot.ResourceNamesFunc = func() []resource.Name {
+			mu.Lock()
+			defer mu.Unlock()
+			callCountNames++
+			return emptyResources
+		}
+		mu.Unlock()
+
+		start := time.Now()
+		dur = 100 * time.Millisecond
```
Comment (cheukt) [inline robot/client/client_test.go:558]: rm

File: robot/client/client_test.go [hunk 5]
```diff
@@ -473,231 +473,194 @@ func TestStatusClient(t *testing.T) {
...
+		client, err := New(
+			context.Background(),
+			listener.Addr().String(),
+			logger,
+			WithRefreshEvery(dur),
+			WithCheckConnectedEvery(dur*2),
+			WithReconnectEvery(dur),
+		)
+		test.That(t, err, test.ShouldBeNil)
+		// block here until ResourceNames is called 6 times
```
Comment (cheukt) [inline robot/client/client_test.go:568]: nit: 7

**Review by cheukt — APPROVED** [review]

**Pull request merged**